crates/matrix-sdk/src/client/mod.rs

```diff
@@ -952,6 +952,48 @@ impl Client {
         self.login_with_token(token.as_str(), device_id, initial_device_display_name).await
     }
 
+    /// Login as appservice virtual user
+    ///
+    /// This function can be used to create an end-to-end-encryption capable
+    /// appservice virtual user.
+    ///
+    /// # Arguments
+    ///
+    /// * `user` - The user that should be logged in to the homeserver.
+    ///
+    /// * `device_id` - A unique id that will be associated with this session.
+    ///   If not given the homeserver will create one. Can be an existing
+    ///   device_id from a previous login call. Note that this should be done
+    ///   only if the client also holds the encryption keys for this device.
+    #[instrument(skip(self, user))]
+    #[cfg(feature = "appservice")]
+    pub async fn appservice_session(
+        &self,
+        user: impl AsRef<str>,
+        device_id: Option<&str>,
+        initial_device_display_name: Option<&str>,
+    ) -> Result<Session> {
+        let homeserver = self.homeserver().await;
+        info!(homeserver = homeserver.as_str(), user = user.as_ref(), "Logging in");
+
+        let login_info = login::v3::LoginInfo::ApplicationService(
+            login::v3::ApplicationService::new(UserIdentifier::UserIdOrLocalpart(user.as_ref())),
+        );
+
+        let request = assign!(login::v3::Request::new(login_info), {
+            device_id: device_id.map(|d| d.into()),
+            initial_device_display_name
+        });
+
+        let response = self.send(request, Some(RequestConfig::short_retry().force_auth())).await?;
```

Comment:
The client doesn't do anything with the response, will this actually create E2EE keys in the end?

The normal `login()` method needs to call `receive_login_response()` so the `UserId` and `DeviceId` get passed to the E2EE state machine. I guess that's not needed because the app-service method `create_and_cache_client()` calls `restore_login()`?

If so, it's not really clear that you need to do additional things to make this work. Perhaps we should clarify the intended usage in the docs.

Reply:
Will add the mention that this needs to be passed to `restore_login`